File: pokemon_card_vault/lib/utils/marketplace_image_log.dart
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;

final Set<String> _recent = <String>{};

void logMarketplaceCardImage({
  required String url,
  String source = 'client',
  String status = 'error',
  String cardId = '',
  String ctId = '',
  String name = '',
  String fallbackUrl = '',
  String route = '',
  Object? error,
}) {
  final trimmed = url.trim();
  if (trimmed.isEmpty) {
    return;
  }
  final key = '$status|$trimmed';
  if (!_recent.add(key)) {
    return;
  }
  if (_recent.length > 80) {
    _recent.clear();
    _recent.add(key);
  }
  debugPrint(
    'marketplace-image $status source=$source cardId=$cardId ctId=$ctId url=$trimmed error=$error',
  );
  if (!kIsWeb && status != 'error') {
    return;
  }
  final uri = Uri.parse('https://api.pokoin.com/api/marketplace-image-log');
  final body = jsonEncode({
    'source': source,
    'status': status,
    'cardId': cardId,
    'ctId': ctId,
    'name': name,
    'url': trimmed,
    'fallbackUrl': fallbackUrl,
    'route': route,
    if (error != null) 'error': '$error',
  });
  http
      .post(
        uri,
        headers: const {'Content-Type': 'application/json'},
        body: body,
      )
      .timeout(const Duration(seconds: 4))
      .catchError((_) => http.Response('', 599));
}

File: pokemon_card_vault/api/marketplace-image-log.js
const {
  listMarketplaceImages,
  recordMarketplaceImage,
} = require('./_marketplace_image_log');

const CORS_HEADERS = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
  'Access-Control-Allow-Headers': 'Content-Type',
  'Access-Control-Max-Age': '86400',
};

const hitsByIp = new Map();

function clientIp(req) {
  const forwarded = String(req.headers['x-forwarded-for'] || '').split(',')[0].trim();
  return forwarded || String(req.socket?.remoteAddress || req.headers['x-real-ip'] || '');
}

function rateLimited(ip) {
  const now = Date.now();
  const windowMs = 60_000;
  const current = hitsByIp.get(ip) || [];
  const fresh = current.filter((stamp) => now - stamp < windowMs);
  if (fresh.length >= 40) {
    hitsByIp.set(ip, fresh);
    return true;
  }
  fresh.push(now);
  hitsByIp.set(ip, fresh);
  return false;
}

module.exports = async function handler(req, res) {
  for (const [key, value] of Object.entries(CORS_HEADERS)) {
    res.setHeader(key, value);
  }
  if (req.method === 'OPTIONS') {
    return res.status(204).end();
  }
  if (req.method === 'GET') {
    const url = new URL(req.url, `https://${req.headers.host || 'pokoin.com'}`);
    const rows = listMarketplaceImages(url.searchParams.get('limit'));
    res.setHeader('Cache-Control', 'no-store');
    return res.status(200).json({ count: rows.length, rows });
  }
  if (req.method !== 'POST') {
    res.setHeader('Allow', 'GET, POST, OPTIONS');
    return res.status(405).json({ error: 'Method not allowed.' });
  }
  if (rateLimited(clientIp(req))) {
    return res.status(429).json({ error: 'Too many image logs.' });
  }
  const body = req.body && typeof req.body === 'object' ? req.body : {};
  const entry = recordMarketplaceImage({
    source: body.source || 'client',
    status: body.status || 'error',
    route: body.route || body.routePath,
    cardId: body.cardId || body.card_id,
    ctId: body.ctId || body.ct_id,
    name: body.name,
    url: body.url || body.imageUrl,
    fallbackUrl: body.fallbackUrl,
    error: body.error,
    sessionId: body.sessionId,
  });
  return res.status(201).json({ ok: true, prefixKind: entry.prefixKind, prefix: entry.prefix });
};

module.exports._test = {
  clientIp,
  rateLimited,
};
